### ISSUE TITLE
Stackable OSS plugins silently fail in cmsd

### ISSUE THREAD

**Issue by bbockelm** [xrootd/xrootd#1730]
I noticed the following in the cmsd logs:

> Config warning: stacked plugins are not supported in this context; directive ignored!
> =====> ofs.osslib ++

It appears that the cmsd silently ignores the stacking, causing the cmsd and xrootd daemons to have differing views of the filesystem.  Could we:

- Enable stacking for cmsd?  It's truly needed here and a few sites have only been functioning by chance.  Right now, I'm just going to give them instructions on how to configure the plugins to do the stacking internally (like what occurred in XRootD 4) which is quite unpleasant.
- Upgrade the stacked plugin failure from a warning to a fatal error?

**Comment by abh3** [comment]
Assuming you are not using forwarding (which we don't really recommend) The fastest way to solve this is to use the "stat" oss plugin and provide whatever view you want from there. The cmsd only relies on Stat() and StatVS() (for writable storage).

**Comment by bbockelm** [comment]
Hi @abh3!

Does the stat oss plugin support stacking?  Using a regular oss plugin works fine here -- I just have to dlopen the underlying plugin and stack myself (which is precisely what we were trying to avoid in V5, no?).

What are your thought on making a stacked plugin failure a fatal error for startup so users aren't surprised?

Brian

**Comment by abh3** [comment]
I agree, we don't want people to manually load stuff. My suggestion here was with the thought that stacking is overkill since all you need is to provide a custom stat() which is what the stat plugin allows you to do. Doesn't seem like that would be complicated to the point of needing to stack anything.

As far as making it a fatal error, well it depends. That is a rather big change in config behavior and we generally like to avoid that kind of thing. I don't have a feel of how many people have tried stacking but I suspect it's only you as no one else has brought this up.

**Comment by bbockelm** [comment]
> I suspect it's only you as no one else has brought this up.

I would be anyone using multiuser + data federation is hitting this.  Once I realized that this was a potential problem, I reached out to Nebraska and noticed they were hitting it as well but had just marked all files world-readable (which is a way around things, I suppose!).

> My suggestion here was with the thought that stacking is overkill since all you need is to provide a custom stat() which is what the stat plugin allows you to do.

In theory, the multiuser plugin can go on top of anything ... in practice, I suppose that it's only stacked on top of the default (POSIX) plugin.  This would require copy/pasting a bunch of code from the default plugin ... but suppose it beats copy/pasting the entire OSS plugin.

If both the stat lib and regular OSS are present, which takes precedence?

**Comment by abh3** [comment]
Since the stat() plugin is a plugin to the standard oss plugin, it simply overrides the default oss behavior. See
https://xrootd.slac.stanford.edu/doc/dev54/ofs_config.htm#_Toc89982421
The doc also mentions an example plugin you can look at.

I agree that the the whole concept here is inverted for instances like yours. I am not against making the plugins for the cmsd stackable but it may not happen on your timescale. Going the stat() plugin route gives you a solution today.

**Comment by xrootd-dev** [comment]
Should be possible but the end result should not have caused any problems 
in practice. That is, if you reply on the end plug-in as that is all that 
really matters. So, I am curious what inthe stacked chain makes this 
problematic. Be aware that the cmsd id not a "user" in any sense of the 
word and assume it has unfethered acces to stat() which is all that it 
cares about.

Andy

**Comment by bbockelm** [comment]
Hi @abh3 -

Not sure I'm entirely following here:

> Should be possible but the end result should not have caused any problems in practice.

The multiuser plugin is necessary for accessing the data from the POSIX filesystem.  It is set to bypass the DACs on the underlying filesystem.

Without it all files in the filesystem must either be world readable (bad) or xrootd readable (requiring complicated sticky group ownership schemes or ACLs).  So indeed, it has caused problems in practice.

We're now having to advise sites to deploy configurations like this:
```
if named xrootd
  ofs.osslib ++ libXrdMultiuser.so
else
  ofs.osslib libXrdMultiuser.so default
fi
```

The second one is where `libXrdMultiuser.so` does the stacking itself.  It's a huge bummer because we've spent years getting people to rip out the plugin-specific stacking and use the XRootD framework's built-in functionality (`++`).

Brian

**Comment by xrootd-dev** [comment]
What I meant by "in practice" is that none of the files need to be world 
readable. The only requirement is that the directories need to be world 
searchable which is not an a particularly big security exposure. Anyway, 
introducing stacking for the cmsd is a messy affair as we would need to 
re-architect the existing code to be more generaly applicable. Certainly, 
this will not be done until 5.6.0.

**Comment by abh3** [comment]
Just getting back to this. The fastest solution may be to use the stat() plugin for the cmsd and then simply doing a setfsuid() for root before the stat() call. As far as I can tell, you can set the fsuid to be root. So, consider this solution (if it actually works).

**Comment by bbockelm** [comment]
Hi Andy,

That sounds fine to me.  The ticket isn’t resolved but let’s close it out until we find a reason to come back to it.

Brian
